### PR TITLE
Add LoRA and Prefix-Tuning as Modeling Options for Improved Memory Efficiency + performance (potentially)

### PR DESCRIPTION
**Pull request by jordiclive**
This PR adds LoRA and prefix-tuning as modelling options (training and sampling code).

Both have shown strong performance and can outperform fine-tuning. They also can protect against the catastrophic forgetting problem which is important for chatbots. They keep the whole language model frozen so they can be distributed freely independent of the base language model.

They also allow much more memory-efficient training as there is no need for the optimizer states of the base model.

Benefits of LoRA
-  Can run only DS 2 with 30B.
- Run 65B without intense CPU usage
- Less overfitting, maintains performance on datasets. 
- Only need to share/push to hub the OS component (small file in Mbs)

— See [Andrej Karpathy (OpenAI)](https://twitter.com/karpathy/status/1649127655122550784?s=20) comment
— See purported [google leak](https://www.semianalysis.com/i/119223672/lora-is-an-incredibly-powerful-technique-we-should-probably-be-paying-more-attention-to)

Implementation Details:
- Explicitly set input_ids as a keyword argument in the sampling code to ensure proper functionality with PEFT `generate.`
-  Manually enable gradients for input to leverage gradient checkpointing effectively, as frozen embeddings would otherwise prevent gradient attachment.
- Include saving code from `pytorch_model.bin` for special tokens. Although these tokens are randomly initialized, they must be stored and saved as an external module since PEFT parameters learn to utilize them. Making them trainable is an option, but it is unlikely to make a significant difference.
- Integrate prefix-tuning, a powerful technique that involves stacking keys and values, by incorporating custom LLama modelling code. (Training speed lower than LoRA in my initial tests and performance similar/worse).

### CODE REVIEW

**Comment by smeyerhot** [comment]
@jordiclive would this PR make it possible to load a PEFT model for inference in the chat?

**Comment by jordiclive** [comment]
This code is currently just for model training and evaluation. But should be trivial to load it for inference, it uses the same HF generate method.

**Comment by jordiclive** [comment]
@andreaskoepf I am going to run a 30B LoRA model just on the sft datasets and will post the sampling report.

File: model/model_training/configs/config.yaml [hunk 1]
```diff
@@ -354,6 +356,78 @@ llama-30b-pretrain:
   num_train_epochs: 1
   save_total_limit: 2
 
+lora-llama-13b:
```
Comment (smeyerhot) [inline model/model_training/configs/config.yaml:359]: Should this be lora-llama-7b?

File: model/model_training/configs/config.yaml [hunk 2]
```diff
@@ -354,6 +356,78 @@ llama-30b-pretrain:
...
+  use_flash_attention: True
+  residual_dropout: 0.0
+  deepspeed_config: configs/zero_config.json
+  peft_model: true
+  peft_type: "lora"
+  use_custom_sampler: true
+
+lora-llama-30b:
+  dtype: fp16
+  log_dir: "llama_log_7b"
```
Comment (smeyerhot) [inline model/model_training/configs/config.yaml:385]: Should these `log_dir`s  be changed to reflect model size?

File: model/model_training/models/peft_modeling.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+@dataclass
+class SaveLoraConfig:
+    dtype: torch.dtype = torch.float16
+    is_reward_model: bool = False
+    quantization: bool = False
+    seq2seqmodel: bool = False
+    freeze_layer: bool = False
+    residual_dropout: float = 0
+    use_flash_attention: bool = False
+    adapter_save_path: str = "adapter_13B_new"
```
Comment (smeyerhot) [inline model/model_training/models/peft_modeling.py:80]: Does this only work on 13B model? Should it be hardcoded?

File: model/model_training/configs/config.yaml (hunk 1 above)
Comment (jordiclive) [reply model/model_training/configs/config.yaml:359]: No sorry, I just didn't bother changing the logging dirs. Will change the log_dir names.

Comment (jordiclive) [reply model/model_training/configs/config.yaml:359]: it is meant to be 13B . you can fit much higher bs. the model_name is a mistake (I had some code to use the 7B model as a tokenizer for all models, and download the weights from HF)

File: model/model_training/models/peft_modeling.py (hunk 3 above)
Comment (jordiclive) [reply model/model_training/models/peft_modeling.py]: it's the output path for a directory with the adapter weights in them. I've changed to just "adapter"

Comment (jordiclive) [reply model/model_training/models/peft_modeling.py]: @smeyerhot I've updated the configs.

**Review by andreaskoepf — APPROVED** [review]
Super great to have this!
